File: validator/validator.js
#!/usr/bin/env node

/**
 * ProtoContext validator — validates a context.txt file against the spec.
 *
 * Usage:
 *   node validator.js <file_or_url>
 *   node validator.js mysite.context.txt
 *   node validator.js https://example.com/context.txt
 */

const fs = require("fs");
const path = require("path");

const MAX_FILE_SIZE = 500 * 1024; // 500KB
const SECTION_WARN_SIZE = 1000;
const REQUIRED_METADATA = new Set(["lang", "version", "updated"]);
const VALID_METADATA = new Set([
  "lang",
  "version",
  "updated",
  "canonical",
  "topics",
  "contact",
  "license",
]);
const DATE_PATTERN = /^\d{4}-\d{2}-\d{2}$/;

function validate(content) {
  const errors = [];
  const warnings = [];

  // Check file size
  const size = Buffer.byteLength(content, "utf-8");
  if (size > MAX_FILE_SIZE) {
    errors.push(`File exceeds 500KB limit (${size.toLocaleString()} bytes)`);
  }

  // Check for HTML/JS
  if (/<\s*(script|style|div|span|html|body|head)\b/i.test(content)) {
    errors.push(
      "File contains HTML/JavaScript — only plain text and basic markdown allowed"
    );
  }

  const lines = content.trim().split("\n");
  if (lines.length === 0) {
    errors.push("File is empty");
    return { valid: errors.length === 0, errors, warnings };
  }

  let idx = 0;

  // Skip leading blank lines
  while (idx < lines.length && lines[idx].trim() === "") idx++;

  // Rule 1: Header
  if (idx >= lines.length || !lines[idx].startsWith("# ")) {
    errors.push("Missing header: file must start with '# Site Name'");
    return { valid: false, errors, warnings };
  }

  const title = lines[idx].slice(2).trim();
  if (!title) errors.push("Header title is empty");
  idx++;

  // Skip blank lines
  while (idx < lines.length && lines[idx].trim() === "") idx++;

  // Description line
  if (idx >= lines.length || !lines[idx].startsWith("> ")) {
    errors.push(
      "Missing description: expected '> One line description' after title"
    );
  } else {
    const desc = lines[idx].slice(2).trim();
    if (!desc) errors.push("Description line is empty");
    if (desc.length > 160) {
      warnings.push(
        `Description is ${desc.length} chars — recommended max is 160`
      );
    }
    idx++;
  }

  // Skip blank lines
  while (idx < lines.length && lines[idx].trim() === "") idx++;

  // Rule 2: Metadata
  const metadata = {};
  while (idx < lines.length && lines[idx].startsWith("@")) {
    const match = lines[idx].match(/^@(\w+):\s*(.+)$/);
    if (match) {
      metadata[match[1]] = match[2].trim();
    } else {
      warnings.push(`Malformed metadata line: '${lines[idx]}'`);
    }
    idx++;
  }

  for (const field of REQUIRED_METADATA) {
    if (!(field in metadata)) {
      errors.push(`Missing required metadata: @${field}`);
    }
  }

  if (metadata.updated && !DATE_PATTERN.test(metadata.updated)) {
    errors.push(
      `@updated must be YYYY-MM-DD format, got: '${metadata.updated}'`
    );
  }

  if (metadata.updated && DATE_PATTERN.test(metadata.updated)) {
    const d = new Date(metadata.updated + "T00:00:00Z");
    if (isNaN(d.getTime())) {
      errors.push(`@updated is not a valid date: '${metadata.updated}'`);
    }
  }

  for (const key of Object.keys(metadata)) {
    if (!VALID_METADATA.has(key)) {
      warnings.push(`Unknown metadata field: @${key}`);
    }
  }

  // Rule 3: Sections
  const sections = [];
  let currentSection = null;

  while (idx < lines.length) {
    const line = lines[idx];
    if (line.startsWith("## section:")) {
      const sectionTitle = line.slice("## section:".length).trim();
      if (!sectionTitle) errors.push("Section title is empty");
      currentSection = { title: sectionTitle, bodyLines: [] };
      sections.push(currentSection);
    } else if (currentSection !== null) {
      currentSection.bodyLines.push(line);
    }
    idx++;
  }

  if (sections.length === 0) {
    errors.push(
      "No sections found — need at least one '## section: Title' block"
    );
  }

  for (const section of sections) {
    const body = section.bodyLines.join("\n").trim();
    if (!body) {
      errors.push(`Section '${section.title}' has no content`);
    }
    if (body.length > SECTION_WARN_SIZE) {
      warnings.push(
        `Section '${section.title}' is ${body.length} chars — recommended max is ~${SECTION_WARN_SIZE} for optimal chunking`
      );
    }
  }

  return { valid: errors.length === 0, errors, warnings };
}

async function validateUrl(url) {
  const https = require("https");
  const http = require("http");
  const mod = url.startsWith("https") ? https : http;

  return new Promise((resolve) => {
    const req = mod.get(
      url,
      { headers: { "User-Agent": "ProtoContext-Validator/1.0" }, timeout: 10000 },
      (res) => {
        const contentType = res.headers["content-type"] || "";
        if (contentType.includes("text/html")) {
          resolve({
            valid: false,
            errors: [
              `URL returned HTML (Content-Type: ${contentType}) — expected text/plain`,
            ],
            warnings: [],
          });
          res.destroy();
          return;
        }

        let data = "";
        res.on("data", (chunk) => (data += chunk));
        res.on("end", () => resolve(validate(data)));
      }
    );

    req.on("error", (e) => {
      resolve({
        valid: false,
        errors: [`Failed to fetch URL: ${e.message}`],
        warnings: [],
      });
    });

    req.on("timeout", () => {
      req.destroy();
      resolve({
        valid: false,
        errors: ["Request timed out after 10 seconds"],
        warnings: [],
      });
    });
  });
}

async function main() {
  const target = process.argv[2];

  if (!target) {
    console.log("Usage: node validator.js <file_or_url>");
    console.log("  node validator.js mysite.context.txt");
    console.log("  node validator.js https://example.com/context.txt");
    process.exit(1);
  }

  let result;

  if (target.startsWith("http://") || target.startsWith("https://")) {
    result = await validateUrl(target);
  } else {
    if (!fs.existsSync(target)) {
      console.log(`INVALID  context.txt\n\n  Errors (1):\n    x File not found: ${target}`);
      process.exit(1);
    }
    const content = fs.readFileSync(target, "utf-8");
    result = validate(content);
  }

  if (result.valid) {
    console.log("VALID  context.txt");
  } else {
    console.log("INVALID  context.txt");
  }

  if (result.errors.length > 0) {
    console.log(`\n  Errors (${result.errors.length}):`);
    for (const err of result.errors) {
      console.log(`    x ${err}`);
    }
  }

  if (result.warnings.length > 0) {
    console.log(`\n  Warnings (${result.warnings.length}):`);
    for (const w of result.warnings) {
      console.log(`    ! ${w}`);
    }
  }

  if (result.valid && result.warnings.length === 0) {
    console.log("  No issues found.");
  }

  process.exit(result.valid ? 0 : 1);
}

main();

File: validator/validator.py
#!/usr/bin/env python3
"""
ProtoContext validator — validates a context.txt file against the spec.

Usage:
    python validator.py <file_or_url>
    python validator.py mysite.context.txt
    python validator.py https://example.com/context.txt
"""

import sys
import re
import os
from datetime import datetime


MAX_FILE_SIZE = 500 * 1024  # 500KB
SECTION_WARN_SIZE = 1000
REQUIRED_METADATA = {"lang", "version", "updated"}
VALID_METADATA = {"lang", "version", "updated", "canonical", "topics", "contact", "license"}
DATE_PATTERN = re.compile(r"^\d{4}-\d{2}-\d{2}$")


class ValidationResult:
    def __init__(self):
        self.errors = []
        self.warnings = []

    def error(self, msg):
        self.errors.append(msg)

    def warn(self, msg):
        self.warnings.append(msg)

    @property
    def valid(self):
        return len(self.errors) == 0


def validate(content: str) -> ValidationResult:
    result = ValidationResult()

    # Check file size
    size = len(content.encode("utf-8"))
    if size > MAX_FILE_SIZE:
        result.error(f"File exceeds 500KB limit ({size:,} bytes)")

    # Check for HTML/JS
    if re.search(r"<\s*(script|style|div|span|html|body|head)\b", content, re.IGNORECASE):
        result.error("File contains HTML/JavaScript — only plain text and basic markdown allowed")

    lines = content.strip().split("\n")
    if not lines:
        result.error("File is empty")
        return result

    idx = 0

    # Skip leading blank lines
    while idx < len(lines) and lines[idx].strip() == "":
        idx += 1

    # Rule 1: Header
    if idx >= len(lines) or not lines[idx].startswith("# "):
        result.error("Missing header: file must start with '# Site Name'")
        return result

    title = lines[idx][2:].strip()
    if not title:
        result.error("Header title is empty")
    idx += 1

    # Skip blank lines
    while idx < len(lines) and lines[idx].strip() == "":
        idx += 1

    # Description line
    if idx >= len(lines) or not lines[idx].startswith("> "):
        result.error("Missing description: expected '> One line description' after title")
    else:
        desc = lines[idx][2:].strip()
        if not desc:
            result.error("Description line is empty")
        if len(desc) > 160:
            result.warn(f"Description is {len(desc)} chars — recommended max is 160")
        idx += 1

    # Skip blank lines
    while idx < len(lines) and lines[idx].strip() == "":
        idx += 1

    # Rule 2: Metadata
    metadata = {}
    while idx < len(lines) and lines[idx].startswith("@"):
        line = lines[idx]
        match = re.match(r"^@(\w+):\s*(.+)$", line)
        if match:
            key, value = match.group(1), match.group(2).strip()
            metadata[key] = value
        else:
            result.warn(f"Malformed metadata line: '{line}'")
        idx += 1

    for field in REQUIRED_METADATA:
        if field not in metadata:
            result.error(f"Missing required metadata: @{field}")

    if "updated" in metadata and not DATE_PATTERN.match(metadata["updated"]):
        result.error(f"@updated must be YYYY-MM-DD format, got: '{metadata['updated']}'")

    if "updated" in metadata and DATE_PATTERN.match(metadata["updated"]):
        try:
            datetime.strptime(metadata["updated"], "%Y-%m-%d")
        except ValueError:
            result.error(f"@updated is not a valid date: '{metadata['updated']}'")

    for key in metadata:
        if key not in VALID_METADATA:
            result.warn(f"Unknown metadata field: @{key}")

    # Rule 3: Sections
    sections = []
    current_section = None

    while idx < len(lines):
        line = lines[idx]
        if line.startswith("## section:"):
            section_title = line[len("## section:"):].strip()
            if not section_title:
                result.error("Section title is empty")
            current_section = {"title": section_title, "body_lines": []}
            sections.append(current_section)
        elif current_section is not None:
            current_section["body_lines"].append(line)
        idx += 1

    if not sections:
        result.error("No sections found — need at least one '## section: Title' block")

    for section in sections:
        body = "\n".join(section["body_lines"]).strip()
        if not body:
            result.error(f"Section '{section['title']}' has no content")
        if len(body) > SECTION_WARN_SIZE:
            result.warn(
                f"Section '{section['title']}' is {len(body)} chars — "
                f"recommended max is ~{SECTION_WARN_SIZE} for optimal chunking"
            )

    return result


def validate_file(path: str) -> ValidationResult:
    if not os.path.exists(path):
        r = ValidationResult()
        r.error(f"File not found: {path}")
        return r

    with open(path, "r", encoding="utf-8") as f:
        content = f.read()

    return validate(content)


def validate_url(url: str) -> ValidationResult:
    import urllib.request
    import urllib.error

    try:
        req = urllib.request.Request(url, headers={"User-Agent": "ProtoContext-Validator/1.0"})
        with urllib.request.urlopen(req, timeout=10) as resp:
            content_type = resp.headers.get("Content-Type", "")
            if "text/html" in content_type:
                r = ValidationResult()
                r.error(f"URL returned HTML (Content-Type: {content_type}) — expected text/plain")
                return r
            content = resp.read().decode("utf-8")
    except urllib.error.URLError as e:
        r = ValidationResult()
        r.error(f"Failed to fetch URL: {e}")
        return r

    return validate(content)


def main():
    if len(sys.argv) < 2:
        print("Usage: python validator.py <file_or_url>")
        print("  python validator.py mysite.context.txt")
        print("  python validator.py https://example.com/context.txt")
        sys.exit(1)

    target = sys.argv[1]

    if target.startswith("http://") or target.startswith("https://"):
        result = validate_url(target)
    else:
        result = validate_file(target)

    if result.valid:
        print(f"VALID  context.txt")
    else:
        print(f"INVALID  context.txt")

    if result.errors:
        print(f"\n  Errors ({len(result.errors)}):")
        for err in result.errors:
            print(f"    x {err}")

    if result.warnings:
        print(f"\n  Warnings ({len(result.warnings)}):")
        for w in result.warnings:
            print(f"    ! {w}")

    if result.valid and not result.warnings:
        print("  No issues found.")

    sys.exit(0 if result.valid else 1)


if __name__ == "__main__":
    main()
